File: c/tree/lowest-common-ancestor-of-a-binary-search-tree[2].c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int find;
int top;
int ptop;
int qtop;
struct TreeNode *parr[5000];
struct TreeNode *qarr[5000];
struct TreeNode *arr[5000];

void scan(struct TreeNode *root, struct TreeNode *node, int flag)
{
    if(root == NULL || find)
        return ;

    arr[++top] = root;
    if(root == node){
        find = 1;
        if(flag){
            ptop = top;
            for(int i = 0; i <= top; ++i)
                parr[i] = arr[i];
        }else{
            qtop = top;
            for(int i = 0; i <= top; ++i)
                qarr[i] = arr[i];
        }
        return;
    }

    scan(root->left, node, flag);
    if(find)
        return;

    scan(root->right, node, flag);
    if(find)
        return;

    --top;
}

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    find = 0;
    top = -1;
    scan(root,p,0);

    find = 0;
    top = -1;
    scan(root,q,1);

    int m = ptop < qtop ? ptop : qtop;
    while(parr[m] != qarr[m])
        m--;

    return parr[m];
}

File: python/tree/lowest-common-ancestor-of-a-binary-search-tree[2].py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lca(self,root,node,flag):
        if root is None or self.find:
            return

        self.rlist.append(root)
        if root == node:
            self.find = 1
            if flag == 0:
                self.plist = self.rlist
            else:
                self.qlist = self.rlist

            return

        self.lca(root.left,node,flag)
        if self.find:
            return

        self.lca(root.right,node,flag)
        if self.find:
            return

        self.rlist.pop()

    def lowestCommonAncestor(self, root, p, q):
        self.rlist = []
        self.find = 0
        self.lca(root,p,0)

        self.rlist = []
        self.find = 0
        self.lca(root,q,1)

        mlen = min(len(self.plist), len(self.qlist)) - 1

        while self.plist[mlen] != self.qlist[mlen]:
            mlen -= 1

        return self.plist[mlen]
